File: HighFive/HighFive/System/Guid.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace System
{
    /// <summary>
    /// The Guid data type which is mapped to the string type in Javascript.
    /// </summary>
    [H5.Immutable]
    public struct Guid : IEquatable<Guid>, IComparable<Guid>, IFormattable
    {
        private const string error1 = "Byte array for GUID must be exactly {0} bytes long";

        private static readonly RegExp Valid = new RegExp("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", "i");
        private static readonly RegExp Split = new RegExp("^(.{8})(.{4})(.{4})(.{4})(.{12})$");
        private static readonly RegExp NonFormat = new RegExp("^[{(]?([0-9a-f]{8})-?([0-9a-f]{4})-?([0-9a-f]{4})-?([0-9a-f]{4})-?([0-9a-f]{12})[)}]?$", "i");
        private static readonly RegExp Replace = new RegExp("-", "g");
        private static readonly Random Rnd = new Random();

        /// <summary>
        /// A read-only instance of the Guid structure whose value is all zeros.
        /// </summary>
        public static readonly Guid Empty = new Guid();

        private int _a;
        private short _b;
        private short _c;
        private byte _d;
        private byte _e;
        private byte _f;
        private byte _g;
        private byte _h;
        private byte _i;
        private byte _j;
        private byte _k;

        /// <summary>
        /// Initializes a new instance of the Guid structure by using the value represented by the specified string.
        /// </summary>
        /// <param name="uuid">A string that contains a GUID</param>
        public Guid(string uuid)
        {
            this = new Guid();

            ParseInternal(uuid, null, true);
        }

        /// <summary>
        /// Initializes a new instance of the Guid structure by using the specified array of bytes.
        /// </summary>
        /// <param name="b">A 16-element byte array containing values with which to initialize the GUID.</param>
        public Guid(byte[] b)
        {
            if (b == null)
            {
                throw new ArgumentNullException("b");
            }

            if (b.Length != 16)
            {
                throw new ArgumentException(string.Format(error1, 16));
            }

            _a = ((int)b[3] << 24) | ((int)b[2] << 16) | ((int)b[1] << 8) | b[0];
            _b = (short)(((int)b[5] << 8) | b[4]);
            _c = (short)(((int)b[7] << 8) | b[6]);
            _d = b[8];
            _e = b[9];
            _f = b[10];
            _g = b[11];
            _h = b[12];
            _i = b[13];
            _j = b[14];
            _k = b[15];
        }

        /// <summary>
        /// Initializes a new instance of the Guid structure by using the specified unsigned integers and bytes.
        /// </summary>
        /// <param name="a">The first 4 bytes of the GUID.</param>
        /// <param name="b">The next 2 bytes of the GUID.</param>
        /// <param name="c">The next 2 bytes of the GUID.</param>
        /// <param name="d">The next byte of the GUID.</param>
        /// <param name="e">The next byte of the GUID.</param>
        /// <param name="f">The next byte of the GUID.</param>
        /// <param name="g">The next byte of the GUID.</param>
        /// <param name="h">The next byte of the GUID.</param>
        /// <param name="i">The next byte of the GUID.</param>
        /// <param name="j">The next byte of the GUID.</param>
        /// <param name="k">The next byte of the GUID.</param>
        [CLSCompliant(false)]
        public Guid(uint a, ushort b, ushort c, byte d, byte e, byte f, byte g, byte h, byte i, byte j, byte k)
        {
            _a = (int)a;
            _b = (short)b;
            _c = (short)c;
            _d = d;
            _e = e;
            _f = f;
            _g = g;
            _h = h;
            _i = i;
            _j = j;
            _k = k;
        }

        /// <summary>
        /// Initializes a new instance of the Guid structure by using the specified integers and byte array.
        /// </summary>
        /// <param name="a">The first 4 bytes of the GUID.</param>
        /// <param name="b">The next 2 bytes of the GUID.</param>
        /// <param name="c">The next 2 bytes of the GUID.</param>
        /// <param name="d">The remaining 8 bytes of the GUID.</param>
        public Guid(int a, short b, short c, byte[] d)
        {
            if (d == null)
            {
                throw new ArgumentNullException("d");
            }

            if (d.Length != 8)
            {
                throw new ArgumentException(string.Format(error1, 8));
            }

            _a = a;
            _b = b;
            _c = c;
            _d = d[0];
            _e = d[1];
            _f = d[2];
            _g = d[3];
            _h = d[4];
            _i = d[5];
            _j = d[6];
            _k = d[7];
        }

        /// <summary>
        /// Initializes a new instance of the Guid structure by using the specified integers and bytes.
        /// </summary>
        /// <param name="a">The first 4 bytes of the GUID.</param>
        /// <param name="b">The next 2 bytes of the GUID.</param>
        /// <param name="c">The next 2 bytes of the GUID.</param>
        /// <param name="d">The next byte of the GUID.</param>
        /// <param name="e">The next byte of the GUID.</param>
        /// <param name="f">The next byte of the GUID.</param>
        /// <param name="g">The next byte of the GUID.</param>
        /// <param name="h">The next byte of the GUID.</param>
        /// <param name="i">The next byte of the GUID.</param>
        /// <param name="j">The next byte of the GUID.</param>
        /// <param name="k">The next byte of the GUID.</param>
        public Guid(int a, short b, short c, byte d, byte e, byte f, byte g, byte h, byte i, byte j, byte k)
        {
            _a = a;
            _b = b;
            _c = c;
            _d = d;
            _e = e;
            _f = f;
            _g = g;
            _h = h;
            _i = i;
            _j = j;
            _k = k;
        }

        /// <summary>
        /// Returns the hash code for this instance.
        /// </summary>
        /// <returns>The hash code for this instance.</returns>
        public override int GetHashCode()
        {
            return this._a ^ (((int)this._b << 16) | (int)(ushort)this._c) ^ (((int)this._f << 24) | this._k);
        }

        /// <summary>
        /// Returns a value indicating whether this instance and a specified Guid object represent the same value.
        /// </summary>
        /// <param name="o">An object to compare to this instance.</param>
        /// <returns>true if o is equal to this instance; otherwise, false.</returns>
        public override bool Equals(Object o)
        {
            if (!(o is Guid))
            {
                return false;
            }

            return this.Equals((Guid)o);
        }

        /// <summary>
        /// Returns a value indicating whether this instance and a specified Guid object represent the same value.
        /// </summary>
        /// <param name="o">An object to compare to this instance.</param>
        /// <returns>true if o is equal to this instance; otherwise, false.</returns>
        public bool Equals(Guid o)
        {
            if ((this._a != o._a) || (this._b != o._b) || (this._c != o._c) || (this._d != o._d)
                || (this._e != o._e) || (this._f != o._f) || (this._g != o._g) || (this._h != o._h)
                || (this._i != o._i) || (this._j != o._j) || (this._k != o._k))
            {
                return false;
            }

            return true;
        }

        /// <summary>
        /// Compares this instance to a specified Guid object and returns an indication of their relative values.
        /// </summary>
        /// <param name="value">An object to compare to this instance.</param>
        /// <returns>A signed number indicating the relative values of this instance and value.</returns>
        public int CompareTo(Guid value)
        {
            return this.ToString().CompareTo(value.ToString());
        }

        /// <summary>
        /// Returns a string representation of the value of this instance in registry format.
        /// </summary>
        /// <returns>The value of this Guid, formatted by using the "D" format specifier.</returns>
        public override string ToString()
        {
            return Format(null);
        }

        /// <summary>
        /// Returns a string representation of the value of this Guid instance, according to the provided format specifier.
        /// </summary>
        /// <param name="format">A single format specifier that indicates how to format the value of this Guid. The format parameter can be "N", "D", "B", "P". If format is null or an empty string (""), "D" is used.</param>
        /// <returns>The value of this Guid, represented as a series of lowercase hexadecimal digits in the specified format.</returns>
        public string ToString(string format)
        {
            return Format(format);
        }

        /// <summary>
        /// Returns a string representation of the value of this instance of the Guid class, according to the provided format specifier and culture-specific format information.
        /// </summary>
        /// <param name="format">A single format specifier that indicates how to format the value of this Guid. The format parameter can be "N", "D", "B", "P". If format is null or an empty string (""), "D" is used.</param>
        /// <param name="formatProvider">(Reserved) An object that supplies culture-specific formatting information.</param>
        /// <returns>The value of this Guid, represented as a series of lowercase hexadecimal digits in the specified format.</returns>
        public string ToString(string format, IFormatProvider formatProvider)
        {
            return Format(format);
        }

        /// <summary>
        /// Returns a 16-element byte array that contains the value of this instance.
        /// </summary>
        /// <returns>A 16-element byte array.</returns>
        public byte[] ToByteArray()
        {
            byte[] g = new byte[16];

            g[0] = (byte)(_a);
            g[1] = (byte)(_a >> 8);
            g[2] = (byte)(_a >> 16);
            g[3] = (byte)(_a >> 24);
            g[4] = (byte)(_b);
            g[5] = (byte)(_b >> 8);
            g[6] = (byte)(_c);
            g[7] = (byte)(_c >> 8);
            g[8] = _d;
            g[9] = _e;
            g[10] = _f;
            g[11] = _g;
            g[12] = _h;
            g[13] = _i;
            g[14] = _j;
            g[15] = _k;

            return g;
        }

        /// <summary>
        /// Converts the string representation of a GUID to the equivalent Guid structure.
        /// </summary>
        /// <param name="input">The string to convert.</param>
        /// <returns>A structure that contains the value that was parsed.</returns>
        public static Guid Parse(string input)
        {
            return ParseExact(input, null);
        }

        /// <summary>
        /// Converts the string representation of a GUID to the equivalent Guid structure, provided that the string is in the specified format.
        /// </summary>
        /// <param name="input">The GUID to convert.</param>
        /// <param name="format">One of the following specifiers that indicates the exact format to use when interpreting input: "N", "D", "B", "P".</param>
        /// <returns></returns>
        public static Guid ParseExact(string input, string format)
        {
            var r = new Guid();
            r.ParseInternal(input, format, true);
            return r;
        }

        /// <summary>
        /// Converts the string representation of a GUID to the equivalent Guid structure.
        /// </summary>
        /// <param name="input">The GUID to convert.</param>
        /// <param name="result">The structure that will contain the parsed value. If the method returns true, result contains a valid Guid. If the method returns false, result equals Guid.Empty.</param>
        /// <returns></returns>
        public static bool TryParse(string input, out Guid result)
        {
            return TryParseExact(input, null, out result);
        }

        /// <summary>
        /// Converts the string representation of a GUID to the equivalent Guid structure, provided that the string is in the specified format.
        /// </summary>
        /// <param name="input">The GUID to convert.</param>
        /// <param name="format">One of the following specifiers that indicates the exact format to use when interpreting input: "N", "D", "B", "P".</param>
        /// <param name="result">The structure that will contain the parsed value. If the method returns true, result contains a valid Guid. If the method returns false, result equals Guid.Empty.</param>
        /// <returns></returns>
        public static bool TryParseExact(string input, string format, out Guid result)
        {
            result = new Guid();
            return result.ParseInternal(input, format, false);
        }

        /// <summary>
        /// Initializes a new instance of the Guid structure.
        /// </summary>
        /// <returns>A new GUID object.</returns>
        public static Guid NewGuid()
        {
            var a = new byte[16];

            Rnd.NextBytes(a);

            a[7] = (byte)(a[7] & 0x0f | 0x40);
            a[8] = (byte)(a[8] & 0xbf | 0x80);

            return new Guid(a);
        }

        /// <summary>
        /// Indicates whether the values of two specified Guid objects are equal.
        /// </summary>
        /// <param name="a">The first object to compare.</param>
        /// <param name="b">The second object to compare.</param>
        /// <returns>true if a and b are equal; otherwise, false.</returns>
        public static bool operator ==(Guid a, Guid b)
        {
            if (Object.ReferenceEquals(a, null))
            {
                return Object.ReferenceEquals(b, null);
            }

            return a.Equals(b);
        }

        /// <summary>
        /// Indicates whether the values of two specified Guid objects are not equal.
        /// </summary>
        /// <param name="a">The first object to compare.</param>
        /// <param name="b">The second object to compare.</param>
        /// <returns>true if a and b are not equal; otherwise, false.</returns>
        public static bool operator !=(Guid a, Guid b)
        {
            return !(a == b);
        }

        private bool ParseInternal(string input, string format, bool check)
        {
            string r = null;

            if (string.IsNullOrEmpty(input))
            {
                if (check)
                {
                    throw new System.ArgumentNullException("input");
                }
                return false;
            }

            if (string.IsNullOrEmpty(format))
            {
                var m = NonFormat.Exec(input);

                if (m != null)
                {
                    List<string> list = new List<string>();
                    for (int i = 1; i <= m.Length; i++)
                    {
                        if (m[i] != null)
                        {
                            list.Add(m[i]);
                        }
                    }

                    r = list.ToArray().Join("-").ToLower();
                }
            }
            else
            {
                format = format.ToUpper();

                var p = false;

                if (format == "N")
                {
                    var m = Split.Exec(input);

                    if (m != null)
                    {
                        List<string> list = new List<string>();
                        for (int i = 1; i <= m.Length; i++)
                        {
                            if (m[i] != null)
                            {
                                list.Add(m[i]);
                            }
                        }

                        p = true;
                        input = list.ToArray().Join("-");
                    }
                }
                else if (format == "B" || format == "P")
                {
                    var b = format == "B" ? new[] { '{', '}' } : new[] { '(', ')' };

                    if ((input[0] == b[0]) && (input[input.Length - 1] == b[1]))
                    {
                        p = true;
                        input = input.Substring(1, input.Length - 2);
                    }
                }
                else
                {
                    p = true;
                }

                if (p && Guid.Valid.Test(input))
                {
                    r = input.ToLower();
                }
            }

            if (r != null)
            {
                FromString(r);
                return true;
            }

            if (check)
            {
                throw new System.FormatException("input is not in a recognized format");
            }

            return false;
        }

        private string Format(string format)
        {
            var s = ToHex((uint)_a, 8) + ToHex((ushort)_b, 4) + ToHex((ushort)_c, 4);
            s = s + (new[] { _d, _e, _f, _g, _h, _i, _j, _k }).Map(ToHex).Join("");

            var m = H5.Script.Write<string[]>("/^(.{8})(.{4})(.{4})(.{4})(.{12})$/.exec(s)");
            string[] list = new string[0];
            for (int i = 1; i < m.Length; i++)
            {
                if (m[i] != null)
                {
                    list.Push(m[i]);
                }
            }
            s = list.Join("-");

            switch (format)
            {
                case "n":
                case "N":
                    return s.ToDynamic().replace(Guid.Replace, "");
                case "b":
                case "B":
                    return '{' + s + '}';

                case "p":
                case "P":
                    return '(' + s + ')';

                default:
                    return s;
            }
        }

        private static string ToHex(uint x, int precision)
        {
            var result = H5.Script.Call<string>("x.toString", 16);
            precision -= result.Length;

            for (var i = 0; i < precision; i++)
            {
                result = "0" + result;
            }

            return result;
        }

        private static string ToHex(byte x)
        {
            var result = H5.Script.Call<string>("x.toString", 16);

            if (result.Length == 1)
            {
                result = "0" + result;
            }

            return result;
        }

        private void FromString(string s)
        {
            if (string.IsNullOrEmpty(s))
            {
                return;
            }

            s = s.ToDynamic().replace(Guid.Replace, "");

            var r = new byte[8];

            _a = (int)uint.Parse(s.Substring(0, 8), 16);
            _b = (short)ushort.Parse(s.Substring(8, 4), 16);
            _c = (short)ushort.Parse(s.Substring(12, 4), 16);
            for (var i = 8; i < 16; i++)
            {
                r[i - 8] = byte.Parse(s.Substring(i * 2, 2), 16);
            }

            _d = r[0];
            _e = r[1];
            _f = r[2];
            _g = r[3];
            _h = r[4];
            _i = r[5];
            _j = r[6];
            _k = r[7];
        }

        private string toJSON()
        {
            return this.ToString();
        }
    }
}

File: Original/Bridge/Bridge/Resources/.generated/System/guid.js
    Bridge.define("System.Guid", {
        inherits: function () { return [System.IEquatable$1(System.Guid),System.IComparable$1(System.Guid),System.IFormattable]; },
        $kind: "struct",
        statics: {
            fields: {
                error1: null,
                Valid: null,
                Split: null,
                NonFormat: null,
                Replace: null,
                Rnd: null,
                Empty: null
            },
            ctors: {
                init: function () {
                    this.Empty = new System.Guid();
                    this.error1 = "Byte array for GUID must be exactly {0} bytes long";
                    this.Valid = new RegExp("^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", "i");
                    this.Split = new RegExp("^(.{8})(.{4})(.{4})(.{4})(.{12})$");
                    this.NonFormat = new RegExp("^[{(]?([0-9a-f]{8})-?([0-9a-f]{4})-?([0-9a-f]{4})-?([0-9a-f]{4})-?([0-9a-f]{12})[)}]?$", "i");
                    this.Replace = new RegExp("-", "g");
                    this.Rnd = new System.Random.ctor();
                }
            },
            methods: {
                Parse: function (input) {
                    return System.Guid.ParseExact(input, null);
                },
                ParseExact: function (input, format) {
                    var r = new System.Guid.ctor();
                    r.ParseInternal(input, format, true);
                    return r;
                },
                TryParse: function (input, result) {
                    return System.Guid.TryParseExact(input, null, result);
                },
                TryParseExact: function (input, format, result) {
                    result.v = new System.Guid.ctor();
                    return result.v.ParseInternal(input, format, false);
                },
                NewGuid: function () {
                    var a = System.Array.init(16, 0, System.Byte);

                    System.Guid.Rnd.NextBytes(a);

                    a[System.Array.index(7, a)] = (a[System.Array.index(7, a)] & 15 | 64) & 255;
                    a[System.Array.index(8, a)] = (a[System.Array.index(8, a)] & 191 | 128) & 255;

                    return new System.Guid.$ctor1(a);
                },
                ToHex$1: function (x, precision) {
                    var result = x.toString(16);
                    precision = (precision - result.length) | 0;

                    for (var i = 0; i < precision; i = (i + 1) | 0) {
                        result = "0" + (result || "");
                    }

                    return result;
                },
                ToHex: function (x) {
                    var result = x.toString(16);

                    if (result.length === 1) {
                        result = "0" + (result || "");
                    }

                    return result;
                },
                op_Equality: function (a, b) {
                    if (Bridge.referenceEquals(a, null)) {
                        return Bridge.referenceEquals(b, null);
                    }

                    return a.equalsT(b);
                },
                op_Inequality: function (a, b) {
                    return !(System.Guid.op_Equality(a, b));
                },
                getDefaultValue: function () { return new System.Guid(); }
            }
        },
        fields: {
            _a: 0,
            _b: 0,
            _c: 0,
            _d: 0,
            _e: 0,
            _f: 0,
            _g: 0,
            _h: 0,
            _i: 0,
            _j: 0,
            _k: 0
        },
        alias: [
            "equalsT", "System$IEquatable$1$System$Guid$equalsT",
            "compareTo", ["System$IComparable$1$System$Guid$compareTo", "System$IComparable$1$compareTo"],
            "format", "System$IFormattable$format"
        ],
        ctors: {
            $ctor4: function (uuid) {
                this.$initialize();
                (new System.Guid.ctor()).$clone(this);

                this.ParseInternal(uuid, null, true);
            },
            $ctor1: function (b) {
                this.$initialize();
                if (b == null) {
                    throw new System.ArgumentNullException.$ctor1("b");
                }

                if (b.length !== 16) {
                    throw new System.ArgumentException.$ctor1(System.String.format(System.Guid.error1, [Bridge.box(16, System.Int32)]));
                }

                this._a = (b[System.Array.index(3, b)] << 24) | (b[System.Array.index(2, b)] << 16) | (b[System.Array.index(1, b)] << 8) | b[System.Array.index(0, b)];
                this._b = Bridge.Int.sxs(((b[System.Array.index(5, b)] << 8) | b[System.Array.index(4, b)]) & 65535);
                this._c = Bridge.Int.sxs(((b[System.Array.index(7, b)] << 8) | b[System.Array.index(6, b)]) & 65535);
                this._d = b[System.Array.index(8, b)];
                this._e = b[System.Array.index(9, b)];
                this._f = b[System.Array.index(10, b)];
                this._g = b[System.Array.index(11, b)];
                this._h = b[System.Array.index(12, b)];
                this._i = b[System.Array.index(13, b)];
                this._j = b[System.Array.index(14, b)];
                this._k = b[System.Array.index(15, b)];
            },
            $ctor5: function (a, b, c, d, e, f, g, h, i, j, k) {
                this.$initialize();
                this._a = a | 0;
                this._b = Bridge.Int.sxs(b & 65535);
                this._c = Bridge.Int.sxs(c & 65535);
                this._d = d;
                this._e = e;
                this._f = f;
                this._g = g;
                this._h = h;
                this._i = i;
                this._j = j;
                this._k = k;
            },
            $ctor3: function (a, b, c, d) {
                this.$initialize();
                if (d == null) {
                    throw new System.ArgumentNullException.$ctor1("d");
                }

                if (d.length !== 8) {
                    throw new System.ArgumentException.$ctor1(System.String.format(System.Guid.error1, [Bridge.box(8, System.Int32)]));
                }

                this._a = a;
                this._b = b;
                this._c = c;
                this._d = d[System.Array.index(0, d)];
                this._e = d[System.Array.index(1, d)];
                this._f = d[System.Array.index(2, d)];
                this._g = d[System.Array.index(3, d)];
                this._h = d[System.Array.index(4, d)];
                this._i = d[System.Array.index(5, d)];
                this._j = d[System.Array.index(6, d)];
                this._k = d[System.Array.index(7, d)];
            },
            $ctor2: function (a, b, c, d, e, f, g, h, i, j, k) {
                this.$initialize();
                this._a = a;
                this._b = b;
                this._c = c;
                this._d = d;
                this._e = e;
                this._f = f;
                this._g = g;
                this._h = h;
                this._i = i;
                this._j = j;
                this._k = k;
            },
            ctor: function () {
                this.$initialize();
            }
        },
        methods: {
            getHashCode: function () {
                return this._a ^ ((this._b << 16) | (this._c & 65535)) ^ ((this._f << 24) | this._k);
            },
            equals: function (o) {
                if (!(Bridge.is(o, System.Guid))) {
                    return false;
                }

                return this.equalsT(System.Nullable.getValue(Bridge.cast(Bridge.unbox(o, System.Guid), System.Guid)));
            },
            equalsT: function (o) {
                if ((this._a !== o._a) || (this._b !== o._b) || (this._c !== o._c) || (this._d !== o._d) || (this._e !== o._e) || (this._f !== o._f) || (this._g !== o._g) || (this._h !== o._h) || (this._i !== o._i) || (this._j !== o._j) || (this._k !== o._k)) {
                    return false;
                }

                return true;
            },
            compareTo: function (value) {
                return System.String.compare(this.toString(), value.toString());
            },
            toString: function () {
                return this.Format(null);
            },
            ToString: function (format) {
                return this.Format(format);
            },
            format: function (format, formatProvider) {
                return this.Format(format);
            },
            ToByteArray: function () {
                var g = System.Array.init(16, 0, System.Byte);

                g[System.Array.index(0, g)] = this._a & 255;
                g[System.Array.index(1, g)] = (this._a >> 8) & 255;
                g[System.Array.index(2, g)] = (this._a >> 16) & 255;
                g[System.Array.index(3, g)] = (this._a >> 24) & 255;
                g[System.Array.index(4, g)] = this._b & 255;
                g[System.Array.index(5, g)] = (this._b >> 8) & 255;
                g[System.Array.index(6, g)] = this._c & 255;
                g[System.Array.index(7, g)] = (this._c >> 8) & 255;
                g[System.Array.index(8, g)] = this._d;
                g[System.Array.index(9, g)] = this._e;
                g[System.Array.index(10, g)] = this._f;
                g[System.Array.index(11, g)] = this._g;
                g[System.Array.index(12, g)] = this._h;
                g[System.Array.index(13, g)] = this._i;
                g[System.Array.index(14, g)] = this._j;
                g[System.Array.index(15, g)] = this._k;

                return g;
            },
            ParseInternal: function (input, format, check) {
                var r = null;

                if (System.String.isNullOrEmpty(input)) {
                    if (check) {
                        throw new System.ArgumentNullException.$ctor1("input");
                    }
                    return false;
                }

                if (System.String.isNullOrEmpty(format)) {
                    var m = System.Guid.NonFormat.exec(input);

                    if (m != null) {
                        var list = new (System.Collections.Generic.List$1(System.String)).ctor();
                        for (var i = 1; i <= m.length; i = (i + 1) | 0) {
                            if (m[i] != null) {
                                list.add(m[i]);
                            }
                        }

                        r = list.ToArray().join("-").toLowerCase();
                    }
                } else {
                    format = format.toUpperCase();

                    var p = false;

                    if (Bridge.referenceEquals(format, "N")) {
                        var m1 = System.Guid.Split.exec(input);

                        if (m1 != null) {
                            var list1 = new (System.Collections.Generic.List$1(System.String)).ctor();
                            for (var i1 = 1; i1 <= m1.length; i1 = (i1 + 1) | 0) {
                                if (m1[i1] != null) {
                                    list1.add(m1[i1]);
                                }
                            }

                            p = true;
                            input = list1.ToArray().join("-");
                        }
                    } else if (Bridge.referenceEquals(format, "B") || Bridge.referenceEquals(format, "P")) {
                        var b = Bridge.referenceEquals(format, "B") ? System.Array.init([123, 125], System.Char) : System.Array.init([40, 41], System.Char);

                        if ((input.charCodeAt(0) === b[System.Array.index(0, b)]) && (input.charCodeAt(((input.length - 1) | 0)) === b[System.Array.index(1, b)])) {
                            p = true;
                            input = input.substr(1, ((input.length - 2) | 0));
                        }
                    } else {
                        p = true;
                    }

                    if (p && System.Guid.Valid.test(input)) {
                        r = input.toLowerCase();
                    }
                }

                if (r != null) {
                    this.FromString(r);
                    return true;
                }

                if (check) {
                    throw new System.FormatException.$ctor1("input is not in a recognized format");
                }

                return false;
            },
            Format: function (format) {
                var s = (System.Guid.ToHex$1((this._a >>> 0), 8) || "") + (System.Guid.ToHex$1((this._b & 65535), 4) || "") + (System.Guid.ToHex$1((this._c & 65535), 4) || "");
                s = (s || "") + ((System.Array.init([this._d, this._e, this._f, this._g, this._h, this._i, this._j, this._k], System.Byte)).map(System.Guid.ToHex).join("") || "");

                var m = /^(.{8})(.{4})(.{4})(.{4})(.{12})$/.exec(s);
                var list = System.Array.init(0, null, System.String);
                for (var i = 1; i < m.length; i = (i + 1) | 0) {
                    if (m[System.Array.index(i, m)] != null) {
                        list.push(m[System.Array.index(i, m)]);
                    }
                }
                s = list.join("-");

                switch (format) {
                    case "n": 
                    case "N": 
                        return s.replace(System.Guid.Replace, "");
                    case "b": 
                    case "B": 
                        return String.fromCharCode(123) + (s || "") + String.fromCharCode(125);
                    case "p": 
                    case "P": 
                        return String.fromCharCode(40) + (s || "") + String.fromCharCode(41);
                    default: 
                        return s;
                }
            },
            FromString: function (s) {
                if (System.String.isNullOrEmpty(s)) {
                    return;
                }

                s = s.replace(System.Guid.Replace, "");

                var r = System.Array.init(8, 0, System.Byte);

                this._a = (System.UInt32.parse(s.substr(0, 8), 16)) | 0;
                this._b = Bridge.Int.sxs((System.UInt16.parse(s.substr(8, 4), 16)) & 65535);
                this._c = Bridge.Int.sxs((System.UInt16.parse(s.substr(12, 4), 16)) & 65535);
                for (var i = 8; i < 16; i = (i + 1) | 0) {
                    r[System.Array.index(((i - 8) | 0), r)] = System.Byte.parse(s.substr(Bridge.Int.mul(i, 2), 2), 16);
                }

                this._d = r[System.Array.index(0, r)];
                this._e = r[System.Array.index(1, r)];
                this._f = r[System.Array.index(2, r)];
                this._g = r[System.Array.index(3, r)];
                this._h = r[System.Array.index(4, r)];
                this._i = r[System.Array.index(5, r)];
                this._j = r[System.Array.index(6, r)];
                this._k = r[System.Array.index(7, r)];
            },
            toJSON: function () {
                return this.toString();
            },
            $clone: function (to) { return this; }
        }
    });
